File: youngjoo00/week_14/BOJ_1753_최단경로.swift
// BOJ_1753_최단경로.swift

// Comparable 프로토콜 채택으로 우선순위 큐에서 구조체의 cost 값을 비교할 수 있게 함
struct EdgeData: Comparable {
    static func < (lhs: EdgeData, rhs: EdgeData) -> Bool {
        return lhs.cost < rhs.cost
    }
    let node: Int
    let cost: Int
}

let input = readLine()!.split(separator: " ").map { Int($0)! }
let V = input[0], E = input[1]
let K = Int(readLine()!)!
let INF = Int.max
var graph = [[EdgeData]](repeating: [], count: V + 1)

for _ in 0..<E {
    let input = readLine()!.split(separator: " ").map { Int($0)! }
    graph[input[0]].append(EdgeData(node: input[1], cost: input[2]))
}

// 가중치 값을 담아둘 배열
var distance = [Int](repeating: INF, count: V + 1)

// 해당 문제는 VE 로는 시간초과가 났음
// 우선순위 큐를 이용하여 VlogE 로 진행
func dijkstra(start: Int) {
    var minHeap = Heap<EdgeData>(comparer: <)
    // 처음 들어온 자기 자신의 가중치 값은 0
    distance[start] = 0
    minHeap.insert(element: EdgeData(node: start, cost: 0))
    
    while !minHeap.isEmpty {
        let now = minHeap.pop()!
        
        // 이미 더 짧은 경로가 존재함을 의미하며, 현재 노드(now)를 통해 인접한 노드들로의 거리를 갱신하는 것은 불필요
        if distance[now.node] < now.cost {
            continue
        }
        
        // 현재 노드와 연결된 노드 확인
        for next in graph[now.node] {
            // 현재노드가 다음 노드로 가는 비용
            let cost = now.cost + next.cost
            
            // 현재 노드를 거쳐서 다음 노드로 가는 비용이 더 적다면 비용 갱신 및 힙에 삽입
            if cost < distance[next.node] {
                distance[next.node] = cost
                minHeap.insert(element: EdgeData(node: next.node, cost: cost))
            }
        }
    }
}

dijkstra(start: K)

distance[1...].forEach { print($0 == INF ? "INF" : $0) }

struct Heap<T: Comparable> {
    private var elements: [T] = []
    private let comparer: (T, T) -> Bool
    
    var isEmpty: Bool {
        return elements.count <= 1
    }
    
    var top: T? {
        return isEmpty ? nil : elements[1]
    }
    
    init(comparer: @escaping (T,T) -> Bool) {
        self.comparer = comparer
    }
    
    mutating func insert(element: T) {
        if elements.isEmpty {
            elements.append(element)
            elements.append(element)
            return
        }
        elements.append(element)
        swimUp(index: elements.count - 1)
    }
    
    mutating private func swimUp(index: Int) {
        var index = index
        while index > 1 && comparer(elements[index], elements[index / 2]) {
            elements.swapAt(index, index / 2)
            index /= 2
        }
    }
    
    mutating func pop() -> T? {
        if elements.count < 2 { return nil }
        elements.swapAt(1, elements.count - 1)
        let deletedElement = elements.removeLast()
        diveDown(index: 1)
        return deletedElement
    }
    
    mutating private func diveDown(index: Int) {
        var swapIndex = index
        var isSwap = false
        let leftIndex = index * 2
        let rightIndex = index * 2 + 1

        if leftIndex < elements.endIndex && comparer(elements[leftIndex], elements[swapIndex]) {
            swapIndex = leftIndex
            isSwap = true
        }
        
        if rightIndex < elements.endIndex && comparer(elements[rightIndex], elements[swapIndex]) {
            swapIndex = rightIndex
            isSwap = true
        }

        if isSwap {
            elements.swapAt(swapIndex, index)
            diveDown(index: swapIndex)
        }
    }
}

File: larcenous/week_14/BOJ_1753_최단경로.py
#heapq 사용
import sys
from heapq import * #우선순위큐로 heap을 사용
input = sys.stdin.readline

V,E = map(int,input().split())
K = int(input().rstrip())
inf = sys.maxsize #거리 초기화를 위한 inf값 설정
graph = [[] for _ in range(V+1)]
for _ in range(E) :
    u,v,w = map(int,input().split())
    graph[u].append((w,v)) #인접 리스트로 그래프 구성 
    #거리 순으로 정렬을 해야하므로, 가중치(거리)를 튜플의 첫번째 원소로 변경 
    #(v,w) -> (w,v)

visited = [False] * (V+1) #방문 여부

D = [inf] * (V+1) #최단거리 리스트
D[K] = 0 #시작노드로의 최단 거리는 0으로 설정

def dijkstra(graph, D, visited, V) :
    min_heap = []
    '''
    #가장 잛은 거리를 가지는 노드를 찾기 위해 모든 노드를 순회
    for i in range(1,V+1) :
        m = -1
        minimum = sys.maxsize
        for j in range(1,V+1) :
            if not visited[j] and D[j] < minimum :
                minimum = D[j]
                m = j
    '''
    heappush(min_heap,(0,K))
    while min_heap : 
        dist, current = heappop(min_heap)
        if visited[current] :
            continue #이미 방문했다면 처음으로 돌아가 최솟값을 다시 pop
        visited[current] = True

        for cost, destination in graph[current] :
            if not visited[destination] :
                if D[current] + cost < D[destination] : #m을 거쳐서 지나가는 경로가 더 짧다면 갱신
                    D[destination] = D[current] + cost
                    heappush(min_heap, (D[destination], destination)) #최소 힙에 (거리, 목적지) 추가

dijkstra(graph, D, visited, V)
for i in range(1,V+1) :
    if D[i] == inf :
        print('INF')
    else :
        print(D[i])